### PR TITLE
A workflow for publishing python package release

### PR DESCRIPTION
**Pull request by wenwei-dev**

### CODE REVIEW

File: .github/workflows/python-publish.yml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        # see https://github.com/conan-io/conan/issues/4322
+        cmake -DCMAKE_C_COMPILER=gcc ..
+
+    - name: Build C API
+      working-directory: ./build
+      run: |
+        set -e
+        make
+        mkdir -p artifacts
+        chrpath -d ./python/*.so  ./c/cargo/*.so
```
Comment (vsbogd) [inline .github/workflows/python-publish.yml:65]: I would include this step into `CMakeLists.txt` files which build the corresponding libraries.

**Review by vsbogd — CHANGES_REQUESTED** [review]
I have tried to install the package from the [test repository](https://test.pypi.org/project/hyperon/0.1.3.4/) and looks like it doesn't contain `.so` libraries.

**Comment by vsbogd** [comment]
Also it is interesting question how could I test it locally. Could you please describe the steps you did for testing?

File: python/setup.py [hunk 2]
```diff
@@ -17,11 +17,18 @@ def run(self):
...
         raise SystemExit(result.returncode)
 
-
 setup(name='hyperon',
-      version='0.1.2',
+      setup_requires=['setuptools_scm'],
+      use_scm_version = {
+          "root": "..",
+          "relative_to": __file__,
+          "local_scheme": "node-and-timestamp"
```
[the remark this answers is not in the document]
Comment (wenwei-dev) [reply python/setup.py:25]: This uses the scm (git tag) to determine the version.

[the remark this answers is not in the document]
Comment (wenwei-dev) [reply python/setup.py:25]: For example, if it is built on git tag v1.0.0, then the version will be v1.0.0

**Comment by vsbogd** [comment]
@wenwei-dev, the problem with incorrect Python version is caused by the fact the GitHub Actions environment has two versions installed:
- `python3-dev` Ubuntu package
- `python` from `actions/python-setup`
By default CMake prefers the system version I am not sure why. But anyway we need to force it find the version we need.
It can be done by specifying `Python3_EXECUTABLE` parameter for the `cmake` command:
```sh
cmake -DPython3_EXECUTABLE=`which python` -DCMAKE_C_COMPILER=gcc -DCMAKE_BUILD_TYPE=Release ..
```
here: https://github.com/wenwei-dev/hyperon-experimental/blob/d7cdc10489c15e1b9bb0d1af84682899723934ca/.github/workflows/python-publish.yml#L85

Also I see that `python-publish.yml` use two different versions in different lines. The `python3` here: https://github.com/wenwei-dev/hyperon-experimental/blob/d7cdc10489c15e1b9bb0d1af84682899723934ca/.github/workflows/python-publish.yml#L77 should be replaced by `python`. This issue probably was copied from `ci.yml`, it also uses `python3` which is incorrect. I will fix it.

**Comment by vsbogd** [comment]
Well, both issues are copied from `ci.yml`. I have fixed them by https://github.com/trueagi-io/hyperon-experimental/pull/285

File: .github/workflows/python-publish.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        tar zcf artifacts.tar.gz artifacts
+
+    - name: Upload shared library artifacts
+      uses: actions/upload-artifact@v3
+      with:
+        name: artifacts.tar.gz
+        path: ./build/artifacts.tar.gz
+
+  publish-python-package:
+    needs: build-ubuntu-2004
```
[the remark this answers is not in the document]
Comment (vsbogd) [reply .github/workflows/python-publish.yml]: implemented in https://github.com/wenwei-dev/hyperon-experimental/pull/2

**Review by vsbogd — APPROVED** [review]
Looks good to me.

**Pull request merged**